x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -331,12 +353,85 @@ public String toString() {
         return this.getClass().getSimpleName() + "@snapshotId=" + snapshotId + " lockFactory=" + lockFactory;
     }
 
+    private void prewarmCache() {
+        if (loadCacheEagerly) {
+            final List<BlobStoreIndexShardSnapshot.FileInfo> cacheFiles = snapshot().indexFiles()
+                .stream()
+                .filter(file -> file.metadata().hashEqualsContents() == false)
+                .filter(file -> isExcludedFromCache(file.physicalName()) == false)
+                .collect(Collectors.toList());
+
+            final Executor executor = threadPool.executor(SEARCHABLE_SNAPSHOTS_THREAD_POOL_NAME);
+            logger.debug("{} warming shard cache for [{}] files", shardId, cacheFiles.size());
+
+            for (BlobStoreIndexShardSnapshot.FileInfo cacheFile : cacheFiles) {
+                final String fileName = cacheFile.physicalName();
+                try {
+                    final IndexInput input = openInput(fileName, CachedBlobContainerIndexInput.CACHE_WARMING_CONTEXT);
+                    assert input instanceof CachedBlobContainerIndexInput : "expected cached index input but got " + input.getClass();
+
+                    final long numberOfParts = cacheFile.numberOfParts();
+                    final CountDown countDown = new CountDown(Math.toIntExact(numberOfParts));
+                    for (long p = 0; p < numberOfParts; p++) {
+                        final int part = Math.toIntExact(p);
+                        executor.execute(new AbstractRunnable() {
```

Comment:
This will queue up all chunks at once on the thread pool. To make this fairer among multiple recoveries, I would prefer something like https://github.com/elastic/elasticsearch/compare/master...ywelsch:para-restores?expand=1

Can be a follow-up though if you prefer.

Reply:
Yes I still have this in mind - I'd prefer to do it in a follow up PR if possible.

Reply:
Could you transfer that to a `// TODO` comment, just so we are sure not to lose it?

Reply:
TODO added in a635c7f064604d74bec4c520580846042aee6310

Reply:
I opened https://github.com/elastic/elasticsearch/pull/55793 for this.